lib/HLSL/DxilValidation.cpp

```diff
@@ -5372,6 +5380,211 @@ static void ValidateEntrySignatures(ValidationContext &ValCtx) {
+    if (F == EntryFn)
+      ValCtx.EmitFnError(EntryFn, ValidationRule::SmIncompatibleOperation);
+    else
+      ValCtx.EmitFnError(EntryFn, ValidationRule::SmIncompatibleCallInEntry);
+
+    // Emit diagnostics for each conflict found in this function.
+    Diagnose(F, conflictMask, ConflictKind::Stage,
+             ValidationRule::SmIncompatibleStage,
+             {ShaderModel::GetKindName(props.shaderKind)});
+    Diagnose(F, conflictMask, ConflictKind::ShaderModel,
+             ValidationRule::SmIncompatibleShaderModel);
+    Diagnose(F, conflictMask, ConflictKind::DerivThreadGroupDim,
+             ValidationRule::SmIncompatibleThreadGroupDim,
+             {std::to_string(props.numThreads[0]),
+              std::to_string(props.numThreads[1]),
+              std::to_string(props.numThreads[2])});
+    Diagnose(F, conflictMask, ConflictKind::DerivInComputeShaderModel,
+             ValidationRule::SmIncompatibleDerivInComputeShaderModel);
+    Diagnose(F, conflictMask, ConflictKind::RequiresGroup,
+             ValidationRule::SmIncompatibleRequiresGroup);
+    Diagnose(F, conflictMask, ConflictKind::DerivLaunch,
+             ValidationRule::SmIncompatibleDerivLaunch,
+             {GetLaunchTypeStr(props.Node.LaunchType)});
+  }
+
+  // Visit function and all functions called by it.
+  // Emit diagnostics for incompatibilities found in a function when no
+  // called functions introduced the conflict.
+  // Return conflict mask for this function.
+  uint32_t Visit(Function *F, uint32_t &remainingMask,
+                 llvm::SmallPtrSet<Function *, 8> &visited, CallGraph &CG) {
+    // Recursive check looks for where a conflict is found and not present
+    // in functions called by the current function.
+    // - When a source is found, emit diagnostics and clear the conflict
+    // flags introduced by this function from the working mask so we don't
+    // report this conflict again.
+    // - When the remainingMask is 0, we are done.
+
+    if (remainingMask == 0)
+      return 0; // Nothing left to search for.
+    if (!visited.insert(F).second)
+      return 0; // Already visited.
+
+    const DxilModule::ShaderCompatInfo *compatInfo =
+        ValCtx.DxilMod.GetCompatInfoForFunction(F);
+    DXASSERT(compatInfo, "otherwise, compat info not computed in module");
+    if (!compatInfo)
+      return 0;
+    uint32_t maskForThisFunction = IdentifyConflict(*compatInfo);
+
+    uint32_t maskForCalls = 0;
+    if (CallGraphNode *CGNode = CG[F]) {
+      for (auto &Call : *CGNode) {
+        Function *called = Call.second->getFunction();
+        if (called->isDeclaration())
+          continue;
+        maskForCalls |= Visit(called, remainingMask, visited, CG);
+        if (remainingMask == 0)
+          return 0; // Nothing left to search for.
+      }
+    }
+
+    // Mask of incompatibilities introduced by this function.
+    uint32_t conflictsIntroduced =
+        remainingMask & maskForThisFunction & ~maskForCalls;
```

Comment:
We are computing `conflictsIntroduced`, the mask of conflicts introduced by this function **that we still need to diagnose**, and that weren't introduced by other functions called by this one.

`remainingMask` has the remaining conflicts that have not yet been diagnosed.  We don't want to diagnose multiple sources of the same conflict under the same call graph, since it would mainly just add noise at that point.  So we remove bits from `remainingMask` when we diagnose that issue, and once there are no remaining conflicts to diagnose, we are done.

Without ANDing in `remainingMask`, we could end up diagnosing the same issue repeatedly.